source/server/config_validation/server.cc

```diff
@@ -0,0 +1,69 @@
+#include "server/config_validation/server.h"
+
+#include "common/json/config_schemas.h"
+
+#include "server/configuration_impl.h"
+
+namespace Server {
+
+ValidationInstance::ValidationInstance(Options& options, HotRestart& restarter,
+                                       Stats::StoreRoot& store,
+                                       Thread::BasicLockable& access_log_lock,
+                                       ComponentFactory& component_factory,
+                                       const LocalInfo::LocalInfo& local_info)
+    : options_(options), restarter_(restarter), stats_store_(store),
+      handler_(Api::ApiPtr{new Api::ValidationImpl(options.fileFlushIntervalMsec())}),
+      local_info_(local_info),
+      access_log_manager_(handler_.api(), handler_.dispatcher(), access_log_lock, store) {
+  try {
+    initialize(options, component_factory);
+  } catch (const EnvoyException& e) {
+    log().critical("error initializing configuration '{}': {}", options.configPath(), e.what());
+    thread_local_.shutdownThread();
+    exit(1);
+  }
+}
+
+void ValidationInstance::initialize(Options& options, ComponentFactory& component_factory) {
+  // See comments on InstanceImpl::initialize() for the overall flow here.
+  //
+  // For validation, we only do a subset of normal server initialization: everything that could fail
+  // on a malformed config (e.g. JSON parsing and all the object construction that follows), but
+  // more importantly nothing with observable effects (e.g. binding to ports or shutting down any
+  // other Envoy process).
+  //
+  // If we get all the way through that stripped-down initialization flow, to the point where we'd
+  // be ready to serve, then the config has passed validation.
+  Json::ObjectPtr config_json = Json::Factory::loadFromFile(options.configPath());
+  config_json->validateSchema(Json::Schema::TOP_LEVEL_CONFIG_SCHEMA);
+  Configuration::InitialImpl initial_config(*config_json);
+
+  for (uint32_t i = 0; i < std::max(1U, options.concurrency()); i++) {
+    workers_.emplace_back(new Worker(thread_local_, options.fileFlushIntervalMsec()));
```

Comment:
Yeah, I was wonder if we can get away with no threading during validation above, but as I look through this method I see there's a fair bit of stuff going on related to this. @mattklein123 Do you think it's possible in Envoy to perform all config validation without bringing up any additional threads?